lom/src/lib/lomtelemetry/client.go

```diff
@@ -0,0 +1,368 @@
+ * Publishers & subscribers may start anytime and can be unware of each other.
+ * This also means that only upon this proxy is started, publishers' data
+ * will reach corresponding subscribers.
+ *
+ * This routine's sole job is to connect publishers & subscribers in full
+ * mesh. This has no special business logic.
+ *
+ * Any one main process can choose to run this proxy. Only one instance
+ * can be run. Any subsequent requests would fail.
+ *
+ * As each channel type has its own independent channels, an independent proxy
+ * is needed per channel type.
+ *
+ * It gets shutdown only upon system shutdown.
+ *
+ * Input:
+ *  chType
+ *      Different channel for different data types. Hence need a proxy
+ *      per type.
+ *
+ * Output:
+ *  None
+ *
+ * Return:
+ *  chClose - Close this channel to stop the proxy
+ *  err - Non nil, in case of failure
+ *
+ */
+func RunPubSubProxy(chType ChannelType_t) (chClose chan<- int, err error) {
+    chCl := make(chan int)
+    err = doRunPubSubProxy(chType, chCl)
+    if err == nil {
+        chClose = chCl
+    }
+    return
+}
+
+/*
+ * Send a request and get channel for receiving response asynchronously
+ *
+ * Input:
+ *  req -   Request to send
+ *          An empty request will shutdown request handler running in background.
+ *          Send empty request for each channel type to ensure closed.
+ *
+ * Output:
+ *  None
+ *
+ * Return:
+ *  chData -    Channel to get response from. Caller to close the channel upon
+ *          receiving response.
+ *  err -   Non nil, in case of failure
+ */
+func SendClientRequest(reqType ChannelType_t, req ClientReq_t) (chData <-chan *ClientRes_t, err error) {
+
+    ch := make(chan *ClientRes_t)
+    if e := processRequest(reqType, req, ch); e != nil {
+        err = cmn.LogError("Failed to process client req err(%v) req(+%v)", e, req)
+        close(ch)
+        ch = nil
+    } else {
+        chData = ch
+    }
+    return
+}
+
+func CloseClientRequest(reqType ChannelType_t) error {
+    return closeRequestChannel(reqType)
+}
+
+/*
+ * Initializes a handler for processing requests.
+ * A handler is for a specific request type.
+ * Only one handler per request can be registered.
+ * Any proc may choose to register a handler.
+ *
+ * Close returned channel chRes to shut this handler.
+ *
+ * Input:
+ *  reqType - Type of request it handles.
+ *
+ * Output:
+ *  None
+ *
+ * Return:
+ *  chDataReq - Input channel to read client requests.
+ *  chRDataes - Output channel to write server's response
+ *  err - Non nil error implies failure
+ */
+
+func RegisterServerReqHandler(reqType ChannelType_t) (chDataReq <-chan ClientReq_t,
+    chDataRes chan<- ServerRes_t, err error) {
+    chReq := make(chan ClientReq_t)
+    chRes := make(chan ServerRes_t)
+
+    if err = initServerRequestHandler(reqType, chReq, chRes); err != nil {
+        /* initServerRequestHandler closes chReq */
+        close(chRes)
```

Comment:
No need to close chReq?

Reply:
channels are always closed by one who has write end,